### PR TITLE
Patch 1

### PR DESCRIPTION
**Pull request by tanaka141**
Added some lines to define LongArray ?

Not tested

### CODE REVIEW

**Comment by udoprog** [comment]
Hm, I don't think these are compatible:
```c++
C->tmp_Section->BlockLight.reset(long_array)
```
All of those section elements are `ByteArray`'s IIUC.

Is this correct? What data in the NBT files require a long arrays nowadays?

**Comment by evildeeds** [comment]
Supporting new format changes should be done in two steps, first lexing/parsing then interpretation. In My opinion this PR should focus om adding support for the long array tag then we can try to figure out what it is used for and if that is relevant for rendering a map.

**Comment by tanaka141** [comment]
Yes, i saw, i didn't look before the diff between 
https://minecraft.gamepedia.com/index.php?title=Chunk_format&oldid=1035105

and the new one
https://minecraft.gamepedia.com/index.php?title=Chunk_format

**Comment by evildeeds** [comment]
Alright; I have now checked, fixed, compiled and tested this and I suggest the following changes before this PR can be accepted
```
diff --git a/src/mc/level.cpp b/src/mc/level.cpp
index 0411ebb..ea4e4bf 100644
--- a/src/mc/level.cpp
+++ b/src/mc/level.cpp
@@ -70,7 +70,7 @@ namespace mc {
 
   struct level_context {
     boost::shared_ptr<Level_Compound> Level;
-    
+
     bool error;
     size_t error_where;
     const char* error_why;
@@ -193,38 +193,38 @@ namespace mc {
     delete byte_array;
   }
 
+  void error_handler(level_context* C, size_t where, const char *why) {
+    C->error = true;
+    C->error_where = where;
+    C->error_why = why;
+  }
+
   void register_long_array(level_context* C, nbt::String name, nbt::LongArray* long_array) {
     if (in_level_section(C))
     {
       if (name.compare("Data") == 0) {
-        C->tmp_Section->Data.reset(long_array);
-        return;
+        std::cout << "warning: register_long_array() - No interpreter available for NBT(Data, LongArray)" << std::endl;
+        //return;
       }
 
       if (name.compare("SkyLight") == 0) {
-        C->tmp_Section->SkyLight.reset(long_array);
-        return;
+        std::cout << "warning: register_long_array() - No interpreter available for NBT(SkyLight, LongArray)" << std::endl;
+        //return;
       }
 
       if (name.compare("BlockLight") == 0) {
-        C->tmp_Section->BlockLight.reset(long_array);
-        return;
+        std::cout << "warning: register_long_array() - No interpreter available for NBT(BlockLight, LongArray)" << std::endl;
+        //return;
       }
 
       if (name.compare("Blocks") == 0) {
-        C->tmp_Section->Blocks.reset(long_array);
-        return;
+        std::cout << "warning: register_long_array() - No interpreter available for NBT(Blocks, LongArray)" << std::endl;
+        //return;
       }
     }
 
     delete long_array;
   }
-  
-  void error_handler(level_context* C, size_t where, const char *why) {
-    C->error = true;
-    C->error_where = where;
-    C->error_why = why;
-  }
 
   level::~level(){
   }
diff --git a/src/nbt/nbt.hpp b/src/nbt/nbt.hpp
index 7a1fc1a..d86ddaa 100644
--- a/src/nbt/nbt.hpp
+++ b/src/nbt/nbt.hpp
@@ -33,7 +33,7 @@ namespace nbt {
     longjmp(exc_env, 1);                                \
   }                                                     \
   } while(0)
-  
+
   #define assert_error_c(exc_env, file, cond, why, cleanup)   \
   do {                                                        \
   if (!(cond)) {                                              \
@@ -411,7 +411,7 @@ namespace nbt {
       inline Byte read_tagType(input_buffer_ptr file) {
         Byte type = read_byte(file);
 
-        nbt_assert_error(exc_env, file, type >= 0 && type <= TAG_Int_Array,
+        nbt_assert_error(exc_env, file, type >= 0 && type <= TAG_Long_Array,
                 "Not a valid tag type");
 
         return type;
@@ -461,7 +461,7 @@ namespace nbt {
       inline void flush_long_array(input_buffer_ptr file) {
         Int length = read_long(file);
 
-        nbt_assert_error(exc_env, file, file->flush(length * sizeof(Int)) != -1,
+        nbt_assert_error(exc_env, file, file->flush(length * sizeof(Long)) != -1,
           "Buffer to short to flush LongArray");
       }
```

**Comment by udoprog** [comment]
So why match against `Data`, `Blocks`, ... at all in `register_long_array`? Aren't they all `ByteArray`s?

**Comment by evildeeds** [comment]
I suggest we keep them for now to verify that the information here https://minecraft.gamepedia.com/index.php?title=Chunk_format is correct since the big warning label suggest otherwise.

**Comment by evildeeds** [comment]
I stopped being lazy and scavenged internet after some random map http://www.minecraftmaps.com/parkour-maps/freaky-parkour that did exhibit a missing tags issue, this PR + fix got me this result instead
```
missing Data
missing BlockLight
missing Blocks
missing Data
missing BlockLight
missing Blocks
missing Data
missing BlockLight
missing Blocks
c10t: /usr/include/boost/smart_ptr/shared_ptr.hpp:734: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = nbt::ByteArray; typename boost::detail::sp_member_access<T>::type = nbt::ByteArray*]: Assertion `px != 0' failed.
Aborted
```

**Comment by tanaka141** [comment]
Big difference between the map format v1.11 and the new one in v1.14.

I think the work to do is on the Compound Structure

**Comment by udoprog** [comment]
fwiw, if you find a level file and you want to figure out the format of it you can use `nbt-inspect` to dump it.

**Comment by tanaka141** [comment]
i have some region file of the new version from a alive internet server.

Where i can find nbt-inspect ?

**Comment by evildeeds** [comment]
Link with Compound Structure notes suggest BlockState (LongArray) instead of Data (ByteArray)
nbt-inspect cannot dump region files by the looks of things
`make nbt-inspect`

**Comment by tanaka141** [comment]
Ok i found that i must make it :D 

after launch all thing, i test it : 
```
./nbt-inspect r.0.1.mca
Unhandled nbt parser error at byte 1: Expected TAG_Compound at root
```
Not really useable, i make another files "region-inspect", and the result is better : 
```
Reading Header
BUFFER SIZE = 131072
CHUNK SIZE: 38099
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 32
    Long(LastUpdate): 4968062
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 0
    Compound(Heightmaps) {
BAD TAG: 0xc
478:Not a valid tag type
BUFFER SIZE = 131072
CHUNK SIZE: 37915
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 32
    Long(LastUpdate): 4968062
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 1
    Compound(Heightmaps) {
BAD TAG: 0xc
478:Not a valid tag type
BUFFER SIZE = 131072
CHUNK SIZE: 37960
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 32
    Long(LastUpdate): 4968062
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 2
    Compound(Heightmaps) {
BAD TAG: 0xc
478:Not a valid tag type
BUFFER SIZE = 131072
CHUNK SIZE: 41376
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 32
    Long(LastUpdate): 4972772
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 3
    Compound(Heightmaps) {
BAD TAG: 0xc
478:Not a valid tag type
```

**Comment by evildeeds** [comment]
> Not really useable, i make another files "region-inspect", and the result is better :
You would think that I knew this utility existed since I have been a contributor to this project for some time, but you would be wrong.
Also there seems to be some type issues still
```
Reading Header
BUFFER SIZE = 131072
CHUNK SIZE: 49964
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 0
    Long(LastUpdate): 890786
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 569676
    Int(xPos): 0
    Compound(Heightmaps) {
1165:Buffer to short to flush LongArray
BUFFER SIZE = 131072
CHUNK SIZE: 40899
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 0
    Long(LastUpdate): 889955
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 451905
    Int(xPos): 1
    Compound(Heightmaps) {
    }
  }
}
```

**Comment by evildeeds** [comment]
in src/nbt/nbt.hpp, function flush_long_array;
the array length is still only an integer. so `Int length = read_int(file);` instead of `Int length = read_long(file);`

**Comment by evildeeds** [comment]
So it looks like this map i got doesn't really contain any map data, unsurprisingly, I'll try grabbing another map.
```
Reading Header
BUFFER SIZE = 131072
CHUNK SIZE: 2852
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): -32
    Long(LastUpdate): 889955
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 0
    Compound(Heightmaps) {
      LongArray(OCEAN_FLOOR): (36 longs)
      LongArray(MOTION_BLOCKING_NO_LEAVES): (36 longs)
      LongArray(MOTION_BLOCKING): (36 longs)
      LongArray(WORLD_SURFACE): (36 longs)
    }
...
    List(Sections, TAG_End, 0): [
    ]
...
Int(DataVersion): 1944
}
```

**Comment by evildeeds** [comment]
Whelp, this is going to be difficult
```
      Compound() {
        LongArray(BlockStates): (256 longs)
        List(Palette, TAG_Compound, 3): [
          Compound() {
            String(Name): minecraft:air
          }
          Compound() {
            String(Name): minecraft:lime_concrete
          }
          Compound() {
            String(Name): minecraft:black_concrete
          }
        ]
        ByteArray(SkyLight): (2048 bytes)
        Byte(Y): 0x6
        ByteArray(BlockLight): (2048 bytes)
      }
```

**Comment by tanaka141** [comment]
Do you want a region file from a alive server ?

**Comment by tanaka141** [comment]
> 
> 
> So it looks like this map i got doesn't really contain any map data, unsurprisingly, I'll try grabbing another map.
> 
> ```
> Reading Header
> BUFFER SIZE = 131072
> CHUNK SIZE: 2852
> Compound() {
>   Compound(Level) {
>     String(Status): full
>     Int(zPos): -32
>     Long(LastUpdate): 889955
>     IntArray(Biomes): (256 ints)
>     Long(InhabitedTime): 0
>     Int(xPos): 0
>     Compound(Heightmaps) {
>       LongArray(OCEAN_FLOOR): (36 longs)
>       LongArray(MOTION_BLOCKING_NO_LEAVES): (36 longs)
>       LongArray(MOTION_BLOCKING): (36 longs)
>       LongArray(WORLD_SURFACE): (36 longs)
>     }
> ...
>     List(Sections, TAG_End, 0): [
>     ]
> ...
> Int(DataVersion): 1944
> }
> ```

How do you get "Heightmaps" readable ? Did you change anything in the Compound structure ?

File: src/mc/level.cpp [hunk 1]
```diff
@@ -193,6 +193,9 @@ namespace mc {
     delete byte_array;
   }
 
+  void register_long_array(level_context* C, nbt::String name, nbt::LongArray* long_array) {
+  }
```
Comment (evildeeds) [inline src/mc/level.cpp]: Default behaviour should be to just delete the array

File: src/mc/region_inspect.cpp [hunk 2]
```diff
@@ -86,6 +86,12 @@ void register_int_array(inspect_context* inspect, nbt::String name, nbt::IntArra
   delete value;
 }
 
+void register_long_array(inspect_context* inspect, nbt::String name, nbt::LongArray* value) {
+  cout << setw(inspect->width) << ""
+       << "LongArray(" << name << "): " << "(" << dec << int(value->length) << " ints)" << endl;
```
Comment (evildeeds) [inline src/mc/region_inspect.cpp]: ints at the end should be longs

File: src/nbt/nbt.hpp [hunk 3]
```diff
@@ -454,6 +457,27 @@ namespace nbt {
         array->length = length;
         register_int_array(context, name, array);
       }
+    
+      inline void flush_long_array(input_buffer_ptr file) {
+        Int length = read_int(file);
+ 
+        nbt_assert_error(exc_env, file, file->flush(length * sizeof(Int)) != -1,
```
Comment (evildeeds) [inline src/nbt/nbt.hpp]: sizeof(Long)

**Comment by evildeeds** [comment]
To make the LongArrays actually print you need to edit the inspect tools, the replace the parsing functions and they do not add a handler for a long array

**Comment by tanaka141** [comment]
Excuse me, i am new with GitHub. I only used it to get some copy from project, never to follow 2 projects with many changes between each other...

**Comment by evildeeds** [comment]
Its okay, here you go
```
diff --git a/src/mc/region_inspect.cpp b/src/mc/region_inspect.cpp
index 8a097b9..7d75e23 100644
--- a/src/mc/region_inspect.cpp
+++ b/src/mc/region_inspect.cpp
@@ -86,6 +86,12 @@ void register_int_array(inspect_context* inspect, nbt::String name, nbt::IntArra
   delete value;
 }
 
+void register_long_array(inspect_context* inspect, nbt::String name, nbt::LongArray* value) {
+  cout << setw(inspect->width) << ""
+       << "LongArray(" << name << "): " << "(" << dec << int(value->length) << " longs)" << endl;
+  delete value;
+}
+
 void error_handler(inspect_context* ctx, size_t where, const char* why) {
   std::cout << where << ":" << why << std::endl;
 }
@@ -115,6 +121,7 @@ int main(int argc, char* argv[]) {
   parser.register_byte = register_byte;
   parser.register_byte_array = register_byte_array;
   parser.register_int_array = register_int_array;
+  parser.register_long_array = register_long_array;
 
   mc::region region(argv[1]);
```

**Comment by tanaka141** [comment]
After all modif, i got this when i use region-inspect

```
Reading Header
BUFFER SIZE = 131072
CHUNK SIZE: 38099
Compound() {
  Compound(Level) {
    String(Status): full
    Int(zPos): 32
    Long(LastUpdate): 4968062
    IntArray(Biomes): (256 ints)
    Long(InhabitedTime): 0
    Int(xPos): 0
    Compound(Heightmaps) {
1144:Not a valid tag type
BUFFER SIZE = 131072
CHUNK SIZE: 37915
Compound() {
  Compound(Level) {
```

**Comment by evildeeds** [comment]
The constant used to validate a tag type also has to be adjusted at https://github.com/udoprog/c10t/blob/001b153f8c767a11411e1b8ab9f4a847298f020c/src/nbt/nbt.hpp#L411
it has to change from TAG_Int_Array to TAG_Long_Array

**Comment by tanaka141** [comment]
What happen now ?

I didn't understand what to do ?

those modifications are merged to the master branch ?

**Comment by evildeeds** [comment]
Either me or @udoprog merge it to master like this

**Comment by tanaka141** [comment]
ok !

I think i saw the solution, but it's more complicated...

**Pull request merged**